### PR TITLE
VersionClients

### PR DESCRIPTION
**Pull request by gdl-koenig**

### CODE REVIEW

File: Gandalan.IDAS.WebApi.Client/RESTRoutinen.cs [hunk 1]
```diff
@@ -288,5 +302,25 @@ public static string GetCurrentMethodName()
...
+                if (!_config.AdditionalHeaders.ContainsKey("api-version"))
+                    _config.AdditionalHeaders.Add("api-version", version.ToString());
+                else
+                    _config.AdditionalHeaders["api-version"] = version.ToString();
+
+                _versionClients.Add(version, HttpClientFactory.GetInstance(_config));
+                return _versionClients.FirstOrDefault(i => i.Key == version).Value;
+            }
+            else
+                return _versionClients.FirstOrDefault(i => i.Key == version).Value;
```
Comment (gdl-reif) [inline Gandalan.IDAS.WebApi.Client/RESTRoutinen.cs:322]: use `_versionClients[version]` instead

File: Gandalan.IDAS.WebApi.Client/RESTRoutinen.cs [hunk 2]
```diff
@@ -288,5 +302,25 @@ public static string GetCurrentMethodName()
...
         }
+
+        public HttpClient GetClientByVersion(string version = null)
+        {
+            if (version == null || _config == null)
+                return _client;
+
+            if (!_versionClients.ContainsKey(version))
+            {
+                if (!_config.AdditionalHeaders.ContainsKey("api-version"))
```
Comment (gdl-reif) [inline Gandalan.IDAS.WebApi.Client/RESTRoutinen.cs:313]: Cleaner IMO: Clone `_config` into a new instance, and always add the header

**Review by gdl-reif — CHANGES_REQUESTED** [review]

File: Gandalan.IDAS.WebApi.Client/BusinessRoutinen/SerienMaterialBerechnenWebRoutinen.cs [hunk 3]
```diff
@@ -29,9 +29,9 @@ public async Task<List<MaterialbedarfDTO>> GetMaterialAsync(Guid serieGuid)
             => await GetAsync<List<MaterialbedarfDTO>>("SerieMaterialbedarfBerechnen?serieGuid=" + serieGuid.ToString());
 
         public async Task<List<MaterialbedarfDTO>> GetOffenerMaterialBedarfAsync(Guid serieGuid)
-            => await GetAsync<List<MaterialbedarfDTO>>("SerieOffenerMaterialbedarf?serieGuid=" + serieGuid.ToString());
+            => await GetAsync<List<MaterialbedarfDTO>>("SerieOffenerMaterialbedarf?serieGuid=" + serieGuid.ToString()/*, version: "2"*/);
         public async Task<List<MaterialbedarfDTO>> GetOffenerMaterialBedarfv1Async(Guid serieGuid)
```
Comment (gdl-reif) [inline Gandalan.IDAS.WebApi.Client/BusinessRoutinen/SerienMaterialBerechnenWebRoutinen.cs:33]: Do it vice versa: keep the original method signature to call v1, and use an explicit v2 in the method name for v2 API call

File: Gandalan.IDAS.WebApi.Client/HttpClientFactory.cs [hunk 4]
```diff
@@ -32,12 +32,25 @@ public class HttpClientConfig
         /// AcceptEncoding Header wird auf GZIP gesetzt.
         /// </summary>
         public bool UseCompression { get; set; }
+
```
Comment (gdl-reif) [inline Gandalan.IDAS.WebApi.Client/HttpClientFactory.cs:35]: Properly implement IClonable

**Review by gdl-reif — CHANGES_REQUESTED** [review]

**Comment by gdl-reif** [comment]
@Saibamen can you do a double check please - I would approve now

File: Gandalan.IDAS.WebApi.Client/RESTRoutinen.cs [hunk 5]
```diff
@@ -61,18 +66,19 @@ public string UserAgent
         /// <param name="url">Relative URL, bezogen auf die BaseUrl</param>
         /// <param name="settings"></param>
         /// <returns>Objektinstanz</returns>
-        public async Task<T> GetAsync<T>(string url, JsonSerializerSettings settings = null)
+        public async Task<T> GetAsync<T>(string url, JsonSerializerSettings settings = null, string version = null)
```
Comment (Saibamen) [inline Gandalan.IDAS.WebApi.Client/RESTRoutinen.cs:69]: Please fix new warnings (`Parameter 'version' has no matching param tag in the XML comment`)

File: Gandalan.IDAS.WebApi.Client/RESTRoutinen.cs [hunk 6]
```diff
@@ -112,19 +119,20 @@ public async Task<byte[]> GetDataAsync(string url)
         /// <param name="data">zu sendendes Objekt</param>
         /// <param name="settings"></param>
         /// <returns>deserialisierte Antwort (i.d.R. sollte das das gespeicherte Objekt in seiner Endfassung sein)</returns>
-        public async Task<T> PostAsync<T>(string url, object data, JsonSerializerSettings settings = null)
+        public async Task<T> PostAsync<T>(string url, object data, JsonSerializerSettings settings = null, string version = null)
```
Comment (Saibamen) [inline Gandalan.IDAS.WebApi.Client/RESTRoutinen.cs:122]: Fix new warning

File: Gandalan.IDAS.WebApi.Client/RESTRoutinen.cs [hunk 7]
```diff
@@ -164,19 +173,20 @@ public async Task<byte[]> PostDataAsync(string url, byte[] data)
         /// <param name="data">zu sendendes Objekt</param>
         /// <param name="settings"></param>
         /// <returns>deserialisierte Antwort (i.d.R. sollte das das gespeicherte Objekt in seiner Endfassung sein)</returns>
-        public async Task<T> PutAsync<T>(string url, object data, JsonSerializerSettings settings = null)
+        public async Task<T> PutAsync<T>(string url, object data, JsonSerializerSettings settings = null, string version = null)
```
Comment (Saibamen) [inline Gandalan.IDAS.WebApi.Client/RESTRoutinen.cs:176]: Fix new warning

File: Gandalan.IDAS.WebApi.Client/RESTRoutinen.cs [hunk 8]
```diff
@@ -212,13 +223,14 @@ public async Task<byte[]> PutDataAsync(string url, byte[] data)
         /// </summary>
         /// <param name="url">Relative URL, bezogen auf die BaseUrl</param>
         /// <returns>Antwort des Servers als String</returns>
-        public async Task DeleteAsync(string url)
+        public async Task DeleteAsync(string url, string version = null)
```
Comment (Saibamen) [inline Gandalan.IDAS.WebApi.Client/RESTRoutinen.cs:226]: Fix new warning

File: Gandalan.IDAS.WebApi.Client/BusinessRoutinen/WebRoutinenBase.cs [hunk 9]
```diff
@@ -313,7 +314,7 @@ public async Task<byte[]> PutDataAsync(string uri, byte[] data, bool skipAuth =
             }
         }
 
-        public async Task DeleteAsync(string uri, bool skipAuth = false)
+        public async Task DeleteAsync(string uri, bool skipAuth = false, string version = null)
```
Comment (Saibamen) [inline Gandalan.IDAS.WebApi.Client/BusinessRoutinen/WebRoutinenBase.cs:317]: A lot of not used optional params (`version`)

**Review by Saibamen — CHANGES_REQUESTED** [review]

**Review by gdl-reif — APPROVED** [review]

**Review by Saibamen — APPROVED** [review]

**Pull request merged**